public/app/features/geo/gazetteer/gazetteer.ts

```diff
@@ -0,0 +1,201 @@
+import { DataFrame, Field, FieldType, KeyValue, toDataFrame } from '@grafana/data';
+import { getBackendSrv } from '@grafana/runtime';
+import { loadWorldmapPoints } from './worldmap';
+import { Geometry, Point } from 'ol/geom';
+import { frameFromGeoJSON } from '../format/geojson';
+import { pointFieldFromLonLat, pointFieldFromGeohash } from '../format/utils';
+import { getCenter } from 'ol/extent';
+
+export interface PlacenameInfo {
+  point: () => Point | undefined; // lon, lat (WGS84)
+  geometry: () => Geometry | undefined;
+  frame?: DataFrame;
+  index?: number;
+}
+
+export interface Gazetteer {
+  path: string;
+  error?: string;
+  find: (key: string) => PlacenameInfo | undefined;
+  examples: (count: number) => string[];
+  frame?: () => DataFrame;
+  count?: number;
+}
+
+// Without knowing the datatype pick a good lookup function
+export function loadGazetteer(path: string, data: any): Gazetteer {
+  // try loading geojson
+  let frame: DataFrame | undefined = undefined;
+
+  if (Array.isArray(data)) {
+    const first = data[0] as any;
+    // Check for legacy worldmap syntax
+    if (first.latitude && first.longitude && (first.key || first.keys)) {
+      return loadWorldmapPoints(path, data);
+    }
+  } else {
+    if (Array.isArray(data?.features) && data?.type === 'FeatureCollection') {
+      frame = frameFromGeoJSON(data);
+    }
+  }
+
+  if (!frame) {
+    try {
+      frame = toDataFrame(data);
+    } catch (ex) {
+      return {
+        path,
+        error: `${ex}`,
+        find: (k) => undefined,
+        examples: (v) => [],
+      };
+    }
+  }
+
+  return frameAsGazetter(frame, { path });
+}
+
+export function frameAsGazetter(frame: DataFrame, opts: { path: string; keys?: string[] }): Gazetteer {
+  const keys: Field[] = [];
+  let geo: Field<Geometry> | undefined = undefined;
+  let lat: Field | undefined = undefined;
+  let lng: Field | undefined = undefined;
+  let geohash: Field | undefined = undefined;
+  let firstString: Field | undefined = undefined;
+  for (const f of frame.fields) {
+    if (f.type === FieldType.geo) {
+      geo = f;
+    }
+    if (!firstString && f.type === FieldType.string) {
+      firstString = f;
+    }
+    if (f.name) {
+      if (opts.keys && opts.keys.includes(f.name)) {
+        keys.push(f);
+      }
+
+      const name = f.name.toUpperCase();
+      switch (name) {
+        case 'LAT':
+        case 'LATITUTE':
+          lat = f;
+          break;
+
+        case 'LON':
+        case 'LNG':
+        case 'LONG':
+        case 'LONGITUE':
+          lng = f;
+          break;
+
+        case 'GEOHASH':
```

Comment:
Might make sense to have these const string values as part of an enum type instead, maybe something like
```
enum Latitude {
  LATITUDE,
  LAT,
}
```

Then for when we are checking we can instead use 3 simple if statements
```
if (name in Latitude) {
  lat = f
}

and two other cases for other enums
```